### DISCUSSION TITLE
kFSEvents Api on Mac OS

### DISCUSSION THREAD

**Discussion by hasansultan92** [libuv/libuv#4237] [General]
Hi everyone, im looking to understand the kFSEvents API for Mac OS.

1. Correct me if I am wrong but I am noticing that libuv is using a "wrapper" so to say with predefined values so that it is capable of compilation without errors using non-apple special compilers? An example I noticed was in the **_src/unix/fsevents.c_** file. My understanding is that libuv is able replicate the values that apple has defined for file events: rename, change, modify, etc. To make libuv compatible with compilers outside of the apple suite, we are using this definition?

<img width="942" alt="image" src="https://github.com/libuv/libuv/assets/52581466/3a3ee99b-df7d-441b-8be9-8bb11bb8bfb6">

2. Additionally, I was wondering if there is a reason that libuv only uses two ways to flag an event in the file system? I see that it is only capable of flagging an event as "rename" or "change". I was able to locate the enum in **_include/uv.h_**
<img width="759" alt="image" src="https://github.com/libuv/libuv/assets/52581466/f1456140-d4d5-4716-8fe1-d69346dad2b0">

**Comment by hasansultan92** [comment]
I have been running some tests with the help of Node js on Mac OS 12.6.3, similar outcomes on ubuntu, where I wanted to watch a dir and run a script if a certain event took place there. I realized that if a file in the dir was renamed, I was expecting to see an event about a file rename and that I should be able to access that file. Unfortunately, when I run an lstat or open on the file, node js would decide to open the file that does not exist (the same file with the original name). I realized that node js is just a messenger and the issue might be originating from libuv or apple itself.

The expected outcome was perhaps along the lines of the following:
![image](https://github.com/libuv/libuv/assets/52581466/3b3e9e6c-0efe-4c27-9d06-06e2193e2515)

**Comment by bnoordhuis** [comment]
> To make libuv compatible with compilers outside of the apple suite, we are using this definition?

Correct. See explanation in libuv/libuv@75c8850c91ffa93fded1c8e2a8e725e00124e4be.

> if there is a reason that libuv only uses two ways to flag an event in the file system?

Libuv abstracts over the file watcher functionality of several operating systems and those were determined to be the greatest common denominator.

**Comment by hasansultan92** [comment]
Thank you for the response @bnoordhuis. Do you think it would be in the best interest of libuv if we were to add more functionality to the file watcher, eg modified, deleted, etc but keep it platform specific. For example: adding ifdef flags in the source code to enable certain events for Mac OS but disable them for Windows and vice versa? Or is that more complicated than it seems in my brain?

**Reply by bnoordhuis** [reply]
> but keep it platform specific

That would go 100% against libuv's primary design goal.

**Reply by hasansultan92** [reply]
hmm I see, do you think I can see some of the docs/references to when you guys made the decision to use the common denominator flags. Just curious to see what OSs are holding this back.

**Comment by hasansultan92** [comment]
@bnoordhuis just to clarify the above statement, I meant to start working with Mac OS and then work towards other OSs to implement the changes. I recall ubuntu having similar outputs, have not gotten the opportunity to test this on Windows yet.

**Comment by hasansultan92** [comment]
Hi everyone, hoping everyone is doing good. I have been working on a platform specific implementation of libuv for the watch_dir functionality. So far i have been doing preliminary testing to see how the platforms are different and what kind of nitty gritty edge cases are arising. My work has been mostly on Ubuntu and Mac OS 

**The linux machine**: _Linux router 5.15.0-89-generic #99~20.04.1-Ubuntu SMP Thu Nov 2 15:16:47 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux_

**The Mac OS machine**: _Darwin MBP.local 21.6.0 Darwin Kernel Version 21.6.0: Mon Dec 19 20:43:09 PST 2022; root:xnu-8020.240.18~2/RELEASE_ARM64_T6000 arm64_


I would like to start off by mentioning the files, functions, enums etc I have altered on my fork - PS I have not made any commits on my fork yet either. Following are the screenshots:

1. **test/test-fs-event.c** _fs_event_cb_dir_multi_file_ <img width="1028" alt="image" src="https://github.com/libuv/libuv/assets/52581466/7465a6c9-b4cd-4a60-b673-2c3af2a4251e">
2. **src/unix/fsevents.c**  _uv__fsevents_event_cb_ <img width="1029" alt="image" src="https://github.com/libuv/libuv/assets/52581466/09a89615-9f84-437c-b97d-e78456719e65">
3. **include/uv.h** _enum uv_fs_event_ 

<img width="487" alt="image" src="https://github.com/libuv/libuv/assets/52581466/8dd00b9d-f5a2-4765-823a-c9b712e45b20">


To drive libuv, on either platforms, I made a simple program that calls the respective handlers to start libuv and call the right functions. If requested I commit them on my fork and we can inspect them there too. Following are the results for the same "scenario"

**Linux** 

- Before I made changes:

<img width="1359" alt="Linux Libuv Test" src="https://github.com/libuv/libuv/assets/52581466/1bf1a1dc-352a-4f17-b5a6-b2af83984032">

- After I made changes

<img width="1359" alt="Linux custom libuv" src="https://github.com/libuv/libuv/assets/52581466/c7862305-3165-43df-8050-5df6bab3bab2">

In either cases, I passed the current test case we have for this functionality

**Mac OS**
<img width="438" alt="image" src="https://github.com/libuv/libuv/assets/52581466/4990218d-48e4-4b52-9a55-228558fc2b09">

- Before I made changes:
<img width="743" alt="image" src="https://github.com/libuv/libuv/assets/52581466/fed09be4-b299-4339-890c-1436326907f5">

- After I made changes:
<img width="501" alt="image" src="https://github.com/libuv/libuv/assets/52581466/e0908650-9bb1-4c0f-beb2-fccb5c371c2a">

- Apple's native API call
<img width="474" alt="image" src="https://github.com/libuv/libuv/assets/52581466/c0dd1395-ef5c-4cbc-97c0-802f942290b2">

Notice Apple's api dumps all the events happening for each "file" event command from the terminal. So when i asked the computer to rename it aka "move" my file, it dumped the whole process it did to complete the task. Depending on who you ask or the usecase, this information may or may not be necessary.

Yes, I am also trying to add support for events like AirDrop on Mac OS. As of writing this comment, Apple does not officially give us an API to interact with AirDrop events like it does for FS events. I might have noticed a trend with AirDropped files where it sets off certain bits and returns a separate range of int value for the `eventflag` so maybe libuv can support that as well in the future, hopefully.

Ideally I want to alter the current test case for `fs_event_watch_dir` because it only really creates and deletes files. We can make it platform specific to see if files are being modified as well.

If this is something we can work with, hopefully, I can continue work on both the platforms and research Windows - I am not well versed with Windows and would very much appreciate some help/support.

**Comment by bnoordhuis** [comment]
Maybe I didn't make myself clear in my previous message but libuv is first and foremost a platform abstraction layer. That means we don't expose platform-specific functionality. If it doesn't work on the majority - preferably: all - of supported platforms, it's not going in.

Concretely, that means something like AirDrop is completely off the table. Even a UV_DELETED event (aside: note the inconsistent tense vis-a-vis UV_CHANGE/UV_RENAME) is probably impossible with a backend like kqueue's EVFILT_VNODE filter.

**Comment by hasansultan92** [comment]
So I was able to implement two new flags UV_CREATE and UV_DELETE tag in 6/9 OSs that Libuv supports (Tier 1 and Tier 3 ones). I was able to test them too. I do not have access to Tier 2 OSs (IBM ones) so I have no implementation for them, what should be the next steps? Should I open a PR for review?
